### DISCUSSION TITLE
How to burn firmware when Mircopython is in mass production

### DISCUSSION THREAD

**Discussion by yfyungit** [micropython/micropython#16224] [General]
Mirco python has bin and.py script file. When the equipment is produced on a large scale, how to quickly burn these two files at once? Are there any tools that can be used?

The hardware in use is ESP8266 ESP32 RP2040 STM32

**Comment by andrewleech** [comment]
In this situation at work we always freeze our python application into the compiled firmware as per:

https://docs.micropython.org/en/latest/reference/manifest.html

**Comment by yfyungit** [comment]
@andrewleech After freezing, update the py file from the network to implement the FOTA function, and there will be multiple code files in flash.

**Comment by andrewleech** [comment]
Ok, you didn't mention you wanted ota firmware updates.
On stm I use mboot and DFU to send encrypted whole firmware updates ota to remote flash, but that only works on stm
There isn't a single standard way of implementing OTA updates on all platforms

**Reply by yfyungit** [reply]
scrubs mentioned using the vfsmmerge tool, which I think is a feasible solution.

**Comment by scruss** [comment]
To manage the multiple code files, there was a similar question last week: [Include some python files in esp32 image without rebuilding · micropython · Discussion #16145](https://github.com/orgs/micropython/discussions/16145#discussioncomment-11137065)

[vfsmerge](https://github.com/Josverl/vfs_merge) was suggested. It may not (yet?) handle all the architectures you need. There are instructions for installing vfsmerge using [poetry](https://github.com/Josverl/vfs_merge/issues/5#issuecomment-2097686699)

**Reply by Josverl** [reply]
Thanks Stewart, 
I was thinking about merge_vfs as well - I have been meaning to include it in mpflash as well to make it simpler to use.
its on [PYPI as well ](https://pypi.org/project/vfs-merge/)

It currently handles LFS filesystems in esp .bin files and any .uf2 files .

For that you can use a process like :

[![](https://mermaid.ink/img/pako:eNp9kt9qwjAUh1_lkCt1WqfsqmzCmHgnu9AxWFskS1MbaJqSP5UivsReYG-yZ9oj7MTOOjZZoSHp9-XwO2n2hKmUk5BkhdqxnGoL63lcAj6L52gpmFZVY3NVwkJouaOaJ561htFstlJOMw71xK_Gg8E4qJpxIP0wSDqtnnbi9LLYqnVmNpLrLY-ibpokHetFn-9vH4WwtuAmMyAkRd5v-anGXBgb-UGLV2cFRr_DXuAKLoZbPjxF-J6jwmg0O-f4FQtuvymczgj1Xj0JpjfBdd-zC1tbpf9vXQ995r-t-N2LgpocRAkZZVbp5lgLU_84X69hd_C4vvf0CMmQYHVJRYo_eO_lmNicSx6TEKcpz6grbEzi8oAqdVatmpKR0GrHh0Qrt81JmNHC4MpVKbV8LuhWU9l95anAPMv2Ch1v0pBUtHxR6uQcvgC-D8Bl?type=png)](https://mermaid.live/edit#pako:eNp9kt9qwjAUh1_lkCt1WqfsqmzCmHgnu9AxWFskS1MbaJqSP5UivsReYG-yZ9oj7MTOOjZZoSHp9-XwO2n2hKmUk5BkhdqxnGoL63lcAj6L52gpmFZVY3NVwkJouaOaJ561htFstlJOMw71xK_Gg8E4qJpxIP0wSDqtnnbi9LLYqnVmNpLrLY-ibpokHetFn-9vH4WwtuAmMyAkRd5v-anGXBgb-UGLV2cFRr_DXuAKLoZbPjxF-J6jwmg0O-f4FQtuvymczgj1Xj0JpjfBdd-zC1tbpf9vXQ995r-t-N2LgpocRAkZZVbp5lgLU_84X69hd_C4vvf0CMmQYHVJRYo_eO_lmNicSx6TEKcpz6grbEzi8oAqdVatmpKR0GrHh0Qrt81JmNHC4MpVKbV8LuhWU9l95anAPMv2Ch1v0pBUtHxR6uQcvgC-D8Bl)


I did some research into FAT32, and it should be possible, but ive not yet implemented it.

stm32 , should/may be possible, but apart from the vfs type (LFS/FAT32) will depend on the tools available for the image file format as there are few cross platform conversion tools.

 Depending on the boards it may/will need an update to a a lookup table after a [manual lookup of the partition table ](https://github.com/Josverl/vfs_merge/blob/main/vfs_merge/portboard_disk.md) or some additional code to simplify that.

**Reply by yfyungit** [reply]
Thank you very much, this seems to be a good solution
